### ISSUE TITLE
Building on Ubuntu 17.10

### ISSUE THREAD

**Issue by kwahoo2** [cdcseacave/openMVS#261]
### Specifications like the version of the project, operating system, and hardware
Linux 4.13.0-16-generic Ubuntu 17.10 x86_64 
AMD Ryzen 7 1700 Eight-Core Processor
Radeon RX 580 Series (POLARIS10 / DRM 3.18.0 / 4.13.0-16-generic, LLVM 5.0.0)
gcc (Ubuntu 7.2.0-8ubuntu3) 7.2.0

### Steps to reproduce the problem
Compilation always fails on Ubuntu 17.10, even with eigen 3.2. 

```
cmake . ../openMVS -DCMAKE_BUILD_TYPE=Release -DVCG_DIR="$main_path/vcglib" -DEIGEN_INCLUDE_DIR="/usr/local/include/eigen3/"

...

-- Found OpenMP_CXX: -fopenmp (found version "1.0") 
-- Found OpenGL: /usr/lib/x86_64-linux-gnu/libGL.so  
CUDA_TOOLKIT_ROOT_DIR not found or specified
-- Could NOT find CUDA (missing: CUDA_TOOLKIT_ROOT_DIR CUDA_NVCC_EXECUTABLE CUDA_INCLUDE_DIRS CUDA_CUDART_LIBRARY) 
-- Can't find CUDA. Continuing without it.
-- WARNING: BREAKPAD was not found: Please specify BREAKPAD directory using BREAKPAD_ROOT env. variable
-- Can't find BreakPad. Continuing without it.
-- Boost version: 1.62.0
-- Found the following Boost libraries:
--   iostreams
--   program_options
--   system
--   serialization
--   regex
-- Eigen 3.2.10 found (include: /usr/local/include/eigen3)
-- OpenCV 3.1.0 found (include: /usr/include/opencv;/usr/include)
-- CGAL 4.9 found (include: /usr//include)
-- VCG  found (include: /home/adi/kompilacje/vcglib)
-- Found required Ceres dependency: Eigen version 3.2.10 in /usr/local/include/eigen3
-- Found Ceres compiled with miniglog substitute for glog, beware this will likely cause problems if glog is later linked.
-- Found Ceres version: 1.13.0 installed in: /usr/local with components: [LAPACK, SuiteSparse, SparseLinearAlgebraLibrary, CXSparse, SchurSpecializations, OpenMP]
-- CERES 1.13.0 found (include: /usr/local/include;/usr/local/include/eigen3;/usr/local/include/ceres/internal/miniglog;/usr/local/include)
-- Configuring done
-- Generating done
-- Build files have been written to: /home/adi/kompilacje/openMVS_build
```

```
make
Scanning dependencies of target Common
[  1%] Building CXX object libs/Common/CMakeFiles/Common.dir/Common.cpp.o
In file included from /usr/local/include/eigen3/Eigen/Core:326:0,
                 from /home/adi/kompilacje/openMVS/libs/Common/Types.h:122,
                 from /home/adi/kompilacje/openMVS/libs/Common/Common.h:176,
                 from /home/adi/kompilacje/openMVS/libs/Common/Common.cpp:12:
/usr/local/include/eigen3/Eigen/src/Core/products/GeneralBlockPanelKernel.h: In member function ‘void Eigen::internal::gebp_kernel<LhsScalar, RhsScalar, Index, mr, nr, ConjugateLhs, ConjugateRhs>::operator()(Eigen::internal::gebp_kernel<LhsScalar, RhsScalar, Index, mr, nr, ConjugateLhs, ConjugateRhs>::ResScalar*, Index, const LhsScalar*, const RhsScalar*, Index, Index, Index, Eigen::internal::gebp_kernel<LhsScalar, RhsScalar, Index, mr, nr, ConjugateLhs, ConjugateRhs>::ResScalar, Index, Index, Index, Index, RhsScalar*)’:
/usr/local/include/eigen3/Eigen/src/Core/products/GeneralBlockPanelKernel.h:574:9: warning: this ‘if’ clause does not guard... [-Wmisleading-indentation]
         if(nr==4) traits.initAcc(C3);
         ^~
/usr/local/include/eigen3/Eigen/src/Core/products/GeneralBlockPanelKernel.h:575:19: note: ...this statement, but the latter is misleadingly indented as if it were guarded by the ‘if’
                   traits.initAcc(C4);
                   ^~~~~~
/usr/local/include/eigen3/Eigen/src/Core/products/GeneralBlockPanelKernel.h:946:9: warning: this ‘if’ clause does not guard... [-Wmisleading-indentation]
         if(nr==4) R3 = ploadu<ResPacket>(r3);
         ^~
/usr/local/include/eigen3/Eigen/src/Core/products/GeneralBlockPanelKernel.h:948:19: note: ...this statement, but the latter is misleadingly indented as if it were guarded by the ‘if’
                   traits.acc(C0, alphav, R0);
                   ^~~~~~
/usr/local/include/eigen3/Eigen/src/Core/products/GeneralBlockPanelKernel.h:951:9: warning: this ‘if’ clause does not guard... [-Wmisleading-indentation]
         if(nr==4) traits.acc(C3, alphav, R3);
         ^~
/usr/local/include/eigen3/Eigen/src/Core/products/GeneralBlockPanelKernel.h:953:19: note: ...this statement, but the latter is misleadingly indented as if it were guarded by the ‘if’
                   pstoreu(r0, R0);
                   ^~~~~~~
In file included from /usr/local/include/eigen3/Eigen/SparseLU:29:0,
                 from /usr/local/include/eigen3/Eigen/Sparse:22,
                 from /home/adi/kompilacje/openMVS/libs/Common/Types.h:124,
                 from /home/adi/kompilacje/openMVS/libs/Common/Common.h:176,
                 from /home/adi/kompilacje/openMVS/libs/Common/Common.cpp:12:
/usr/local/include/eigen3/Eigen/src/SparseLU/SparseLU_gemm_kernel.h: In function ‘void Eigen::internal::sparselu_gemm(Index, Index, Index, const Scalar*, Index, const Scalar*, Index, Scalar*, Index)’:
/usr/local/include/eigen3/Eigen/src/SparseLU/SparseLU_gemm_kernel.h:78:9: warning: this ‘if’ clause does not guard... [-Wmisleading-indentation]
         if(RK==4) b30 = pset1<Packet>(Bc0[3]);
         ^~
/usr/local/include/eigen3/Eigen/src/SparseLU/SparseLU_gemm_kernel.h:79:19: note: ...this statement, but the latter is misleadingly indented as if it were guarded by the ‘if’
                   b01 = pset1<Packet>(Bc1[0]);
                   ^~~
In file included from /home/adi/kompilacje/openMVS/libs/Common/Types.h:369:0,
                 from /home/adi/kompilacje/openMVS/libs/Common/Common.h:176,
                 from /home/adi/kompilacje/openMVS/libs/Common/Common.cpp:12:
/home/adi/kompilacje/openMVS/libs/Common/Hash.h: In static member function ‘static SEACAVE::cHashTable<Type>::Key SEACAVE::cHashTable<Type>::HashKey(const SEACAVE::String&)’:
/home/adi/kompilacje/openMVS/libs/Common/Hash.h:246:87: error: no matching function for call to ‘SEACAVE::cHashTable<Type>::HashKey(const char*, std::__cxx11::basic_string<char>::size_type)’
  static inline Key HashKey(const String& str) { return HashKey(str.c_str(), str.size()); }
                                                                                       ^
/home/adi/kompilacje/openMVS/libs/Common/Hash.h:244:20: note: candidate: static SEACAVE::cHashTable<Type>::Key SEACAVE::cHashTable<Type>::HashKey(const uint8_t*, UINT) <near match>
  static inline Key HashKey(const uint8_t* data, UINT size) { return HashKeyFNV(data, size); }
                    ^~~~~~~
/home/adi/kompilacje/openMVS/libs/Common/Hash.h:244:20: note:   conversion of argument 1 would be ill-formed:
/home/adi/kompilacje/openMVS/libs/Common/Hash.h:245:20: note: candidate: static SEACAVE::cHashTable<Type>::Key SEACAVE::cHashTable<Type>::HashKey(LPCTSTR)
  static inline Key HashKey(LPCTSTR sz) { return HashKey((const uint8_t*)sz, _tcslen(sz)); }
                    ^~~~~~~
/home/adi/kompilacje/openMVS/libs/Common/Hash.h:245:20: note:   candidate expects 1 argument, 2 provided
/home/adi/kompilacje/openMVS/libs/Common/Hash.h:246:20: note: candidate: static SEACAVE::cHashTable<Type>::Key SEACAVE::cHashTable<Type>::HashKey(const SEACAVE::String&)
  static inline Key HashKey(const String& str) { return HashKey(str.c_str(), str.size()); }
                    ^~~~~~~
/home/adi/kompilacje/openMVS/libs/Common/Hash.h:246:20: note:   candidate expects 1 argument, 2 provided
In file included from /home/adi/kompilacje/openMVS/libs/Common/Types.h:377:0,
                 from /home/adi/kompilacje/openMVS/libs/Common/Common.h:176,
                 from /home/adi/kompilacje/openMVS/libs/Common/Common.cpp:12:
/home/adi/kompilacje/openMVS/libs/Common/HalfFloat.h: In static member function ‘static short int SEACAVE::hfloat::fromFloat(float)’:
/home/adi/kompilacje/openMVS/libs/Common/HalfFloat.h:62:29: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
   const int& fltInt32((int&)fv);
                             ^~
/home/adi/kompilacje/openMVS/libs/Common/HalfFloat.h: In static member function ‘static float SEACAVE::hfloat::toFloat(short int)’:
/home/adi/kompilacje/openMVS/libs/Common/HalfFloat.h:70:18: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
   return (float&)fltInt32;
                  ^~~~~~~~
In file included from /home/adi/kompilacje/openMVS/libs/Common/Types.h:425:0,
                 from /home/adi/kompilacje/openMVS/libs/Common/Common.h:176,
                 from /home/adi/kompilacje/openMVS/libs/Common/Common.cpp:12:
/home/adi/kompilacje/openMVS/libs/Common/SML.h: In destructor ‘SEACAVE::SMLVALUE_TYPE::~SMLVALUE_TYPE()’:
/home/adi/kompilacje/openMVS/libs/Common/SML.h:29:28: warning: deleting ‘void*’ is undefined [-Wdelete-incomplete]
  ~SMLVALUE_TYPE() { delete data; }
                            ^~~~
In file included from /home/adi/kompilacje/openMVS/libs/Common/Common.h:176:0,
                 from /home/adi/kompilacje/openMVS/libs/Common/Common.cpp:12:
/home/adi/kompilacje/openMVS/libs/Common/Types.h: In function ‘float SEACAVE::cbrt5(float)’:
/home/adi/kompilacje/openMVS/libs/Common/Types.h:735:8: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
  (int&)f = (((int&)f-(127<<23))/3+(127<<23));
        ^
/home/adi/kompilacje/openMVS/libs/Common/Types.h:735:20: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
  (int&)f = (((int&)f-(127<<23))/3+(127<<23));
                    ^
/home/adi/kompilacje/openMVS/libs/Common/Types.h: In member function ‘int SEACAVE::Float::Floor2Int() const’:
/home/adi/kompilacje/openMVS/libs/Common/Types.h:1176:94: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
  inline int  Floor2Int() const    { const double d((double)v.f + FLOOR_FIX); return (*(int*)&d) >> 16; }
                                                                                              ^
/home/adi/kompilacje/openMVS/libs/Common/Types.h: In member function ‘int SEACAVE::Float::Ceil2Int() const’:
/home/adi/kompilacje/openMVS/libs/Common/Types.h:1179:99: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
  inline int  Ceil2Int() const    { const double d((double)v.f + 1.0 + FLOOR_FIX); return (*(int*)&d) >> 16; }
                                                                                                   ^
In file included from /home/adi/kompilacje/openMVS/libs/Common/Types.h:2730:0,
                 from /home/adi/kompilacje/openMVS/libs/Common/Common.h:176,
                 from /home/adi/kompilacje/openMVS/libs/Common/Common.cpp:12:
/home/adi/kompilacje/openMVS/libs/Common/Types.inl: In member function ‘void SEACAVE::TImage<TYPE>::toGray(SEACAVE::TImage<T>&, int, bool) const’:
/home/adi/kompilacje/openMVS/libs/Common/Types.inl:2065:33: error: cannot call member function ‘int cv::Mat::channels() const’ without object
  const int scn(cv::Mat::channels());
                ~~~~~~~~~~~~~~~~~^~
libs/Common/CMakeFiles/Common.dir/build.make:62: polecenia dla obiektu 'libs/Common/CMakeFiles/Common.dir/Common.cpp.o' nie powiodły się
make[2]: *** [libs/Common/CMakeFiles/Common.dir/Common.cpp.o] Błąd 1
CMakeFiles/Makefile2:144: polecenia dla obiektu 'libs/Common/CMakeFiles/Common.dir/all' nie powiodły się
make[1]: *** [libs/Common/CMakeFiles/Common.dir/all] Błąd 2
Makefile:129: polecenia dla obiektu 'all' nie powiodły się
make: *** [all] Błąd 2
```

Meanwhile, building on 16.04 (I put one in virtualbox) works 100% fine.

**Comment by cdcseacave** [comment]
I do not have a Ubuntu 17.10 machine. Feel free to add a PR with the fix. Thanks.

**Comment by himanshusagar** [comment]
For all those facing this issue, By default Ubuntu 17.10 has opencv 3.x, you need to install opencv 2.x [(HERE)](https://launchpad.net/ubuntu/artful/amd64/libopencv-dev/2.4.9.1+dfsg1-2) to solve it. However this will affect other packages like blender.

**Comment by ThomasEgi** [comment]
If you want to avoid the old opencv version you have to fix the errors.
The first error occuring is in Hash.h line 246. You can fix it by static_casting to the required types.
static inline Key HashKey(const String& str) { return HashKey(static_cast<uint8_t*>(str.c_str()), static_cast<UINT> (str.size())); }

The second error is in Types.inl line 2065, i managed to compile it by deleting the cv::Mat:: part so it tries to get the actual method of the object instead of the class method.
const int scn(channels());

Those two changes do make the compiler happy on 17.10, it even compiles with the eigen version shipped ( 3.3.4-3 ). However i'm not very experienced in c++ especially not with inline stuff so it might break stuff in other places. It did successfully compile on my machine, and it did run and produced a dense point cloud for me, it exited with error code 11 tho so it didn't finish correctly i guess. 

If there is a better solution i'd love to hear, other than this it seems to run nicely with opencv 3.

**Comment by cdcseacave** [comment]
could you please make a PR containing these changes?

**Comment by ThomasEgi** [comment]
https://github.com/cdcseacave/openMVS/pull/291 is up

**Comment by kwahoo2** [comment]
I can confirm that #291 works for me, with default OpenCV and Eigen shipped by Ubuntu 17.10.
